Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -192,266 +89,700 @@ void LoadSettings()
+						nameToDockedInfoMap[dockedShipName].lastDockedSolar = CreateID(ini.get_value_string(1));
+
+						ci.dockedShipList.emplace_back(dockedShipName);
+						dockedCount++;
+					}
+				}
+				if (doLoad)
+				{
+					nameToCarrierInfoMap[carrierName] = ci;
+					carrierCount++;
+				}
+				else
+				{
+					for (const auto& dockedShipName : ci.dockedShipList)
+					{
+						MoveOfflineShipToLastDockedSolar(dockedShipName);
+					}
+				}
+			}
+		}
+		ini.close();
+	}
 	ConPrint(L"DockingModules: Loaded %u equipment\n", dockingModAmount);
+	ConPrint(L"DockingModules: Found %u ships docked on %u carriers\n", dockedCount, carrierCount);
 	ConPrint(L"DockingModules: Allowing ships below the cargo capacity of %i to dock\n", cargoCapacityLimit);
 }
 
-void __stdcall BaseExit(uint iBaseID, uint iClientID)
+void SaveData()
 {
-	returncode = DEFAULT_RETURNCODE;
+	while (true)
+	{
+		std::this_thread::sleep_for(std::chrono::minutes(1));
 
-	mobiledockClients[iClientID].iDockingModulesInstalled = GetInstalledModules(iClientID);
+		if (nameToCarrierInfoMap.empty())
+		{
+			continue;
+		}
 
-	// Normalize the docking modules available, with the number of people currently docked
-	mobiledockClients[iClientID].iDockingModulesAvailable = (mobiledockClients[iClientID].iDockingModulesInstalled - mobiledockClients[iClientID].mapDockedShips.size());
+		char datapath[MAX_PATH];
+		GetUserDataPath(datapath);
+		string path = string(datapath) + R"(\Accts\MultiPlayer\docking_module\mobile_docking.ini)";
+		FILE* file = fopen(path.c_str(), "w");
+		if (file)
+		{
+			std::lock_guard<std::mutex> saveLock(saveMutex);
+			for (const auto& ci : nameToCarrierInfoMap)
+			{
+				if (ci.second.dockedShipList.empty())
+				{
+					continue;
+				}
+				fprintf(file, "[CarrierData]\n");
+				fprintf(file, "carrier = %ls\n", ci.first.c_str());
+				fprintf(file, "lastLogin = %u\n", (uint)ci.second.lastCarrierLogin);
+				for (const wstring& dockedName : ci.second.dockedShipList)
+				{
+					const auto& lastSolarInfo = Universe::get_base(nameToDockedInfoMap[dockedName.c_str()].lastDockedSolar);
+					fprintf(file, "docked = %ls, %s\n", dockedName.c_str(), lastSolarInfo->cNickname);
+				}
+			}
 
-	// If this is a ship which is currently docked, clear the market
-	if (mobiledockClients[iClientID].mobileDocked)
-	{
-		SendResetMarketOverride(iClientID);
+			fclose(file);
+		}
+		else
+		{
+			ConPrint(L"ERROR MobileDock failed to open the player data file!\n");
+			AddLog("ERROR MobileDock failed to open the player data file!\n");
+		}
 	}
-
 }
 
-// Temporary storage for client data to be handled in LaunchPosHook.
-CLIENT_DATA undockingShip;
+wstring GetLastBaseName(uint client)
+{
+	const auto& dockedInfo = idToDockedInfoMap[client];
+	const auto& baseInfo = Universe::get_base(dockedInfo->lastDockedSolar);
+	const auto& sysInfo = Universe::get_system(baseInfo->iSystemID);
+	wstring baseName;
+	wstring& sysName = HkGetWStringFromIDS(sysInfo->strid_name);

```

Comment:
I'm slightly nervous about the ability for these HkGetWStringFromIDS calls to blow up in our faces if a mod developer screws up referential integrity (which they do all the time) and names a base after some non-existent infocard

Reply:
HkGetWStringFromIDS() already defaults in case of a nonexistent IDS (I believe to "Object Unknown"?)

Reply:
It does?

Reply:
Per discussion - it's L"" and the BHkLoadStringDLLs call was added to mobiledocking's DllMain